File: Wiki.Web/Controllers/CharactersController.vb
Imports System
Imports System.Collections.Generic
Imports System.Data
Imports System.Data.Entity
Imports System.Linq
Imports System.Net
Imports System.Web
Imports System.Web.Mvc
Imports Wiki.Data.Entity.Context
Imports Wiki.Domain

Namespace Controllers
    Public Class CharactersController
        Inherits System.Web.Mvc.Controller

        Private db As New CharacterDbContext

        ' GET: Characters
        Function Index() As ActionResult
            Return View(db.Characters.ToList())
        End Function

        ' GET: Characters/Details/5
        Function Details(ByVal id As Integer?) As ActionResult
            If IsNothing(id) Then
                Return New HttpStatusCodeResult(HttpStatusCode.BadRequest)
            End If
            Dim character As Character = db.Characters.Find(id)
            If IsNothing(character) Then
                Return HttpNotFound()
            End If
            Return View(character)
        End Function

        ' GET: Characters/Create
        Function Create() As ActionResult
            Return View()
        End Function

        ' POST: Characters/Create
        'To protect from overposting attacks, enable the specific properties you want to bind to, for 
        'more details see https://go.microsoft.com/fwlink/?LinkId=317598.
        <HttpPost()>
        <ValidateAntiForgeryToken()>
        Function Create(<Bind(Include:="Id,Name,Description,Type,Origin")> ByVal character As Character) As ActionResult
            If ModelState.IsValid Then
                db.Characters.Add(character)
                db.SaveChanges()
                Return RedirectToAction("Index")
            End If
            Return View(character)
        End Function

        ' GET: Characters/Edit/5
        Function Edit(ByVal id As Integer?) As ActionResult
            If IsNothing(id) Then
                Return New HttpStatusCodeResult(HttpStatusCode.BadRequest)
            End If
            Dim character As Character = db.Characters.Find(id)
            If IsNothing(character) Then
                Return HttpNotFound()
            End If
            Return View(character)
        End Function

        ' POST: Characters/Edit/5
        'To protect from overposting attacks, enable the specific properties you want to bind to, for 
        'more details see https://go.microsoft.com/fwlink/?LinkId=317598.
        <HttpPost()>
        <ValidateAntiForgeryToken()>
        Function Edit(<Bind(Include:="Id,Name,Description,Type,Origin")> ByVal character As Character) As ActionResult
            If ModelState.IsValid Then
                db.Entry(character).State = EntityState.Modified
                db.SaveChanges()
                Return RedirectToAction("Index")
            End If
            Return View(character)
        End Function

        ' GET: Characters/Delete/5
        Function Delete(ByVal id As Integer?) As ActionResult
            If IsNothing(id) Then
                Return New HttpStatusCodeResult(HttpStatusCode.BadRequest)
            End If
            Dim character As Character = db.Characters.Find(id)
            If IsNothing(character) Then
                Return HttpNotFound()
            End If
            Return View(character)
        End Function

        ' POST: Characters/Delete/5
        <HttpPost()>
        <ActionName("Delete")>
        <ValidateAntiForgeryToken()>
        Function DeleteConfirmed(ByVal id As Integer) As ActionResult
            Dim character As Character = db.Characters.Find(id)
            db.Characters.Remove(character)
            db.SaveChanges()
            Return RedirectToAction("Index")
        End Function

        Protected Overrides Sub Dispose(ByVal disposing As Boolean)
            If (disposing) Then
                db.Dispose()
            End If
            MyBase.Dispose(disposing)
        End Sub
    End Class
End Namespace

File: Wiki.Web/Controllers/CharactersController.cs
using AutoMapper;
using System;
using System.Collections.Generic;
using System.Data;
using System.Data.Entity;
using System.Linq;
using System.Net;
using System.Web;
using System.Web.Mvc;
using Wiki.Business;
using Wiki.Data.Entity.Context;
using Wiki.Repositories.Common;
using Wiki.Repositories.Entity;
using Wiki.Web.ViewModels.Character;
using Wiki.Web.ViewModels.CharacterType;

namespace Wiki.Web.Controllers
{
    [Authorize]
    public class CharactersController : Controller
    {
        private IBaseRepository<Character, int> repositoryCharacters = new CharactersRepository(new WikiDbContext());
        private IBaseRepository<CharacterType, int> repositoryCharacterTypes = new CharacterTypesRepository(new WikiDbContext());

        // GET: Characters
        public ActionResult Index()
        {

            return View(Mapper.Map<List<Character>, List<CharacterIndexViewModel>>(repositoryCharacters.List()));
        }
        public ActionResult FilterByName(string name)
        {
            List<Character> characters = repositoryCharacters.List();
            if (name != null)
            {
                characters = characters.Where(a => a.Name.Contains(name)).ToList();
            }

            return Json(Mapper.Map<List<Character>, List<CharacterIndexViewModel>>(characters), JsonRequestBehavior.AllowGet);
        }



        // GET: Characters/Details/5
        public ActionResult Details(int? id)
        {
            if (id == null)
            {
                return new HttpStatusCodeResult(HttpStatusCode.BadRequest);
            }
            Character character = repositoryCharacters.Find(id.Value);
            if (character == null)
            {
                return HttpNotFound();
            }
            return View(Mapper.Map<Character, CharacterIndexViewModel>(character));
        }

        // GET: Characters/Create
        public ActionResult Create()
        {
            ViewBag.IdType = new SelectList(Mapper.Map<List<CharacterType>, List<CharacterTypeIndexViewModel>>(repositoryCharacterTypes.List()), "Id", "Name");
            return View();
        }

        // POST: Characters/Create
        // To protect from overposting attacks, enable the specific properties you want to bind to, for 
        // more details see https://go.microsoft.com/fwlink/?LinkId=317598.
        [HttpPost]
        [ValidateAntiForgeryToken]
        public ActionResult Create([Bind(Include = "Id,Name,Description,IdType,Origin")] CharacterViewModel characterViewModel)
        {
            if (ModelState.IsValid)
            {
                repositoryCharacters.Add(Mapper.Map<CharacterViewModel, Character>(characterViewModel));
                return RedirectToAction("Index");
            }

            ViewBag.IdType = new SelectList(Mapper.Map<List<CharacterType>, List<CharacterTypeIndexViewModel>>(repositoryCharacterTypes.List()), "Id", "Name", characterViewModel.IdType);
            return View(characterViewModel);
        }

        // GET: Characters/Edit/5
        public ActionResult Edit(int? id)
        {
            if (id == null)
            {
                return new HttpStatusCodeResult(HttpStatusCode.BadRequest);
            }
            Character character = repositoryCharacters.Find(id.Value);
            if (character == null)
            {
                return HttpNotFound();
            }
            ViewBag.IdType = new SelectList(Mapper.Map<List<CharacterType>, List<CharacterTypeIndexViewModel>>(repositoryCharacterTypes.List()), "Id", "Name", character.IdType);
            return View(Mapper.Map<Character, CharacterViewModel>(character));
        }

        // POST: Characters/Edit/5
        // To protect from overposting attacks, enable the specific properties you want to bind to, for 
        // more details see https://go.microsoft.com/fwlink/?LinkId=317598.
        [HttpPost]
        [ValidateAntiForgeryToken]
        public ActionResult Edit([Bind(Include = "Id,Name,Description,IdType,Origin")] CharacterViewModel characterViewModel)
        {
            if (ModelState.IsValid)
            {
                repositoryCharacters.Edit(Mapper.Map<CharacterViewModel, Character>(characterViewModel));
                return RedirectToAction("Index");
            }
            ViewBag.IdType = new SelectList(Mapper.Map<List<CharacterType>, List<CharacterTypeIndexViewModel>>(repositoryCharacterTypes.List()), "Id", "Name", characterViewModel.IdType);
            return View(characterViewModel);
        }

        // GET: Characters/Delete/5
        public ActionResult Delete(int? id)
        {
            if (id == null)
            {
                return new HttpStatusCodeResult(HttpStatusCode.BadRequest);
            }
            Character character = repositoryCharacters.Find(id.Value);
            if (character == null)
            {
                return HttpNotFound();
            }
            return View(Mapper.Map<Character, CharacterIndexViewModel>(character));
        }

        // POST: Characters/Delete/5
        [HttpPost, ActionName("Delete")]
        [ValidateAntiForgeryToken]
        public ActionResult DeleteConfirmed(int id)
        {
            repositoryCharacters.RemoveById(id);
            return RedirectToAction("Index");
        }
    }
}
